src/librustc_traits/chalk/db.rs

```diff
@@ -175,9 +175,39 @@ impl<'tcx> chalk_solve::RustIrDatabase<RustInterner<'tcx>> for RustIrDatabase<'t
 
     fn fn_def_datum(
         &self,
-        _fn_def_id: chalk_ir::FnDefId<RustInterner<'tcx>>,
+        fn_def_id: chalk_ir::FnDefId<RustInterner<'tcx>>,
     ) -> Arc<chalk_solve::rust_ir::FnDefDatum<RustInterner<'tcx>>> {
-        unimplemented!()
+        let def_id = match fn_def_id.0 {
+            RustDefId::FnDef(def_id) => def_id,
+            _ => bug!("Did not use `FnDef` variant when expecting FnDef."),
+        };
+        let bound_vars = bound_vars_for_item(self.tcx, def_id);
+        let binders = binders_for(&self.interner, bound_vars);
+
+        let predicates = self.tcx.predicates_defined_on(def_id).predicates;
+        let where_clauses: Vec<_> = predicates
+            .into_iter()
+            .map(|(wc, _)| wc.subst(self.tcx, &bound_vars))
+            .filter_map(|wc| LowerInto::<Option<chalk_ir::QuantifiedWhereClause<RustInterner<'tcx>>>>::lower_into(wc, &self.interner)).collect();
+
+        let sig = self.tcx.fn_sig(def_id);
+        // FIXME(chalk): Why does this have a Binder
```

Comment:
Because of late-bound regions -- I overlooked that, we should fix chalk to have a binder as well. For a function like

```rust
fn foo<'a>(x: &'a u32) { }
```

the `'a` is considered "late-bound" and doesn't show up in the generics of `foo`, but rather in its signature.

Reply:
We should probably not invoke `skip_binder` here but rather something like `no_bound_vars().expect("FIXME: late-bound fn parameters not supported in chalk")`

Reply:
That's weird. So do lifetimes never show up in the generics of functions?

Reply:
Ugh so any functions that take `&self` (e.g. in `inherent_impl`) are broken